### ISSUE TITLE
Proposal: Flatten package hierarchy

### ISSUE THREAD

**Issue by twpayne** [kubernetes-sigs/e2e-framework#150]
Firstly, thank you for this software. We're using it to write new end-to-end tests for [Cilium](https://github.com/cilium/cilium).

While the API is not yet finalized (the latest version is 0.0.7 as I'm writing this), would you consider making some changes to the package hierarchy to make this module easier to use and more idiomatic Go?

Currently, the module consists of 13 commonly-used subpackages (`klient`, `klient/conf`, `klient/decoder`, `klient/k8s`, `klient/k8s/resources`, `klient/k8s/watcher`, `klient/wait`, `klient/wait/conditions`, `pkg/env`, `pkg/envconf`, `pkg/envfuncs`, `pkg/features`, and `pkg/flags`), each with a few functions and types. Idiomatic Go, (e.g. [the standard library's `os` package](https://pkg.go.dev/os)) instead favor few subpackages each with many functions and types. Would you consider reducing the number of subpackages? This would have the following benefits:
* Finding and navigating the documentation becomes much easier, as there is much more in a single package. With the current structure I have 13 open tabs in Chrome just for sigs.k8s.io/e2e-framework, and understanding the type of any individual function involves navigating through multiple packages.
* Fine-grained packages require a lot of import statements, and each import statement squats on a file-level identifier that then cannot be used for other purposes. For example, I would like to be able to write `resources := envconf.New().Resources()` but cannot use the variable name `resources` as it is already squatted by the import of `sigs.k8s.io/e2e-framework/klient/k8s/resources`.

Ideally this change to be more idiomatic Go would happen while e2e-framework's API is still unstable, as these are backwards-incompatible changes.

If you are willing to consider such a change, I can put together a proposal with precise details or create a PR.

cc @gandro @tklauser @willfindlay

**Comment by vladimirvivien** [comment]
@twpayne thanks for the proposal. 

As with all software endeavors, there are reasons for evolution of a project structure (some organic while others planned). The design is package-first, meaning a package is the entry point to a functionality thus reducing name clashes, i.e. I can do `envconf.New()` to create a new environment conf,  in a flatten package approach it would be `flatpackage.NewConf()` or `flatpackage.NewWatcher`, etc. A flatten structure reduces the namespace capacity of the package, forcing elaborate function naming.

Furthermore, if the gold standard for idiomatic Go package organization  is the standard library, I propose you also look at  package `crypto` (`math`, and others). It has a structured layout (similar to what's discussed here) where each cryptographic algorithm  implementation is encapsulated in its own sub-package. I will agree that subpackage make it easy to squat variable names, as you pointed out, but that is a side effect of Go and can easily be fixed with a simpler variable name change. 

**Provide a simple example**
Having said all of that, it is possible I am missing the entire point that you are making. So, please reply with proposed package structure (could be just a list of them) and code example how it would make writing tests simpler.

**Why I think this could be a problem**
* It would break tests for folks who already are using the framework (there is a quite a few) 
* I am not convinced it is a Go problem (idiomatic or otherwise)
* I don't think it would make writing tests easier
* As the project scales in size (more functionalities), a flatten approach will be refactored into a structured layout (as seen in many projects).

**Comment by twpayne** [comment]
Thank you for considering this. I do realize that such changes will not be backwards compatible, which is why I propose that this should be done while the project is still young.

For idiomatic Go, there are guidelines on package names [in Effective Go](https://go.dev/doc/effective_go#package-names) and [in Go's Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#package-names). Examples of how the current package structure is different to the guidelines include the current `./pkg/flags` subpackage which has a naming conflict with the standard library's `flags` package. An example current package name that is confusing is `./klient/k8s` which gets imported as `k8s`. The use of `k8s` here strongly suggests that a `k8s.Object` is something from a `k8s.io` module but really it's something specific to the `sigs.k8s.io/e2e-framework` module.

Here is an initial sketch of some proposed renames/refactoring:

## `klient` package and subpackages

### Types

| Current package and type name | Proposed package and type name |
| - | - |
| `klient.Client` | `klient.Client` |
| `klient.k8s.resources.Resources` | `klient.Resources` |
| `klient.k8s.watcher.EventHandlerFuncs` | `klient.Watcher` |
| `klient.k8s.wait.For` | `klient.WaitFor` |
| `klient.k8s.wait.conditions.Condition` | `klient.WaitCondition` |
| `klient.k8s.Object` | `klient.K8sObject` |
| `klient.k8s.ObjectList` | `klient.K8sObjectList` |

## Constructors

| Current package and name | Proposed package and name |
| - | - |
| `klient.New` | `klient.NewClient` |
| `klient.conf.New` | `klient.NewRESTConfig` |
| `klient.conf.NewWithContextName` | `klient.NewRESTConfigWithContextName` |
| `klient.k8s.resources.New` | `klient.NewResources` |

## `pkg` package and subpackages

Rename to `e2e` and refactor:

### Types

| Current package and name | Proposed package and name |
| - | - |
| `pkg.env.Environment` | `e2e.Environment` |
| `pkg.env.Func` | `e2e.EnvFunc` |
| `pkg.env.FeatureFunc` | `e2e.FeatureFunc` |
| `pkg.envconf.Config` | `e2e.EnvConfig` |
| `pkg.features.FeatureBuilder` | `e2e.FeatureBuilder` |
| `pkg.flags.EnvFlags` | `e2e.EnvFlags` |

### Constructors

| Current package and name | Proposed package and name |
| - | - |
| `pkg.env.New` | `e2e.NewEnvironment` |
| `pkg.envconf.New` | `e2e.NewEnvConfig` |

### Functions

| Current package and name | Proposed package and name |
| - | - |
| `pkg.envfuncs.CreateNamespace` | `e2e.CreateNamespace` |

## Impact

If we land such a change, users will have to change a few function names and imports, which I believe can be done automatically by `go fmt`. Here's the diff of the changes required to the wait for resources example:

```diff
diff --git a/examples/wait_for_resources/wait_test.go b/examples/wait_for_resources/wait_test.go
index ef244c4..607c704 100644
--- a/examples/wait_for_resources/wait_test.go
+++ b/examples/wait_for_resources/wait_test.go
@@ -25,17 +25,14 @@ import (
        v1 "k8s.io/api/core/v1"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/labels"
-       "sigs.k8s.io/e2e-framework/klient/k8s"
-       "sigs.k8s.io/e2e-framework/klient/k8s/resources"
-       "sigs.k8s.io/e2e-framework/klient/wait"
-       "sigs.k8s.io/e2e-framework/klient/wait/conditions"
-       "sigs.k8s.io/e2e-framework/pkg/envconf"
-       "sigs.k8s.io/e2e-framework/pkg/features"
+       "sigs.k8s.io/e2e-framework/e2e"
+       "sigs.k8s.io/e2e-framework/klient"
 )
 
 func TestWaitForResources(t *testing.T) {
-       depFeature := features.New("appsv1/deployment").WithLabel("env", "dev").
-               Setup(func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+       depFeature := e2e.NewFeatureBuilder("appsv1/deployment").WithLabel("env", "dev").
+               Setup(func(ctx context.Context, t *testing.T, cfg *e2e.EnvConfig) context.Context {
                        // create a deployment
                        deployment := newDeployment(cfg.Namespace(), "test-deployment", 10)
                        client, err := cfg.NewClient()
@@ -47,7 +44,7 @@ func TestWaitForResources(t *testing.T) {
                        }
                        return ctx
                }).
-               Assess("deployment >=50% available", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+               Assess("deployment >=50% available", func(ctx context.Context, t *testing.T, cfg *e2e.EnvConfig) context.Context {
                        client, err := cfg.NewClient()
                        if err != nil {
                                t.Fatal(err)
@@ -56,17 +53,17 @@ func TestWaitForResources(t *testing.T) {
                                ObjectMeta: metav1.ObjectMeta{Name: "test-deployment", Namespace: cfg.Namespace()},
                        }
                        // wait for the deployment to become at least 50%
-                       err = wait.For(conditions.New(client.Resources()).ResourceMatch(&dep, func(object k8s.Object) bool {
+                       err = klient.WaitFor(klient.NewCondition(client.Resources()).ResourceMatch(&dep, func(object klient.K8sObject) bool {
                                d := object.(*appsv1.Deployment)
                                return float64(d.Status.ReadyReplicas)/float64(*d.Spec.Replicas) >= 0.50
-                       }), wait.WithTimeout(time.Minute*1))
+                       }), client.WaitWithTimeout(time.Minute*1))
                        if err != nil {
                                t.Fatal(err)
                        }
                        t.Logf("deployment availability: %.2f%%", float64(dep.Status.ReadyReplicas)/float64(*dep.Spec.Replicas)*100)
                        return ctx
                }).
-               Assess("deployment available", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+               Assess("deployment available", func(ctx context.Context, t *testing.T, cfg *e2e.EnvConfig) context.Context {
                        client, err := cfg.NewClient()
                        if err != nil {
                                t.Fatal(err)
@@ -75,20 +72,20 @@ func TestWaitForResources(t *testing.T) {
                                ObjectMeta: metav1.ObjectMeta{Name: "test-deployment", Namespace: cfg.Namespace()},
                        }
                        // wait for the deployment to finish becoming available
-                       err = wait.For(conditions.New(client.Resources()).DeploymentConditionMatch(&dep, appsv1.DeploymentAvailable, v1.ConditionTrue), wait.WithTimeout(time.Minute*1))
+                       err = klient.WaitFor(klient.NewCondition(client.Resources()).DeploymentConditionMatch(&dep, appsv1.DeploymentAvailable, v1.ConditionTrue), klient.WaitWithTimeout(time.Minute*1))
                        if err != nil {
                                t.Fatal(err)
                        }
                        return ctx
                }).
-               Assess("deployment pod garbage collection", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+               Assess("deployment pod garbage collection", func(ctx context.Context, t *testing.T, cfg *e2e.EnvConfig) context.Context {
                        client, err := cfg.NewClient()
                        if err != nil {
                                t.Fatal(err)
                        }
                        // get list of pods
                        var pods v1.PodList
-                       err = client.Resources(cfg.Namespace()).List(context.TODO(), &pods, resources.WithLabelSelector(labels.FormatLabels(map[string]string{"app": "wait-for-resources"})))
+                       err = client.Resources(cfg.Namespace()).List(context.TODO(), &pods, klient.ResourcesWithLabelSelector(labels.FormatLabels(map[string]string{"app": "wait-for-resources"})))
                        if err != nil {
                                t.Fatal(err)
                        }
@@ -101,7 +98,7 @@ func TestWaitForResources(t *testing.T) {
                                t.Fatal(err)
                        }
                        // wait for the deployment pods to be deleted
-                       err = wait.For(conditions.New(client.Resources()).ResourcesDeleted(&pods), wait.WithTimeout(time.Minute*1))
+                       err = klient.WaitFor(klient.NewCondition(client.Resources()).ResourcesDeleted(&pods), klient.WaitWithTimeout(time.Minute*1))
                        if err != nil {
                                t.Fatal(err)
                        }
```

Note that:
* There are far fewer imports, freeing up multiple useful identifiers like `resources`, `wait`, `conditions`, and `features` to be used in code.
* There are only two packages imported for the e2e framework, which means users will see all the functionality provided by e2e-framework on just two pages on pkg.go.dev (as opposed to the current 22 packages in e2e-framework), making discoverability and use much easier.

**Comment by harshanarayana** [comment]
@vladimirvivien  @twpayne 

> Maybe one thing we could do is create a PR to play around with the ideas presented here.

I took this as a request and went ahead and did a draft refactoring account for both of your suggestions and a few changes of my own that I think might be useful down the line. 

https://github.com/harshanarayana/e2e-framework/tree/feature/GIT-150/code-reorg

I took inspiration from `k/k` 's `test/e2e/framework` to restructure some of these and added a few additional packages that were initially missing earlier. Around setting up a pluggable providers that can be hooked into the e2e framework similar to kind.  I am sure there are still a good bit of things that can certainly be cleaned up and done better in the draft that I have. I will spend a little more time this weekend and refine that private branch a bit more. 

> NOTE: I have not fully run the test cases. So please be kind if the tests fail ;)
